client.go

```diff
@@ -0,0 +1,198 @@
+	UnmarshalError func(resp *http.Response) error
+}
+
+var DefaultErrorUnmarshaler = ErrorUnmarshaler(new(RemoteError))
+
+// Call invokes the endpoint implied by the given params,
+// which should be of the form accepted by the ArgT
+// argument to a function passed to Handle, and
+// unmarshals the response into the given response parameter,
+// which should be a pointer to the response value.
+//
+// If resp is nil, the response will be ignored if the
+// request was successful.
+//
+// If resp is of type **http.Response, instead of unmarshaling
+// into it, its element will be set to the returned HTTP
+// response directly and the caller is responsible for
+// closing its Body field.
+//
+// Any error that c.UnmarshalError or c.Doer returns will not
+// have its cause masked.
+func (c *Client) Call(baseURL string, params, resp interface{}) error {
+	rt, err := getRequestType(reflect.TypeOf(params))
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	if rt.method == "" {
+		return errgo.Newf("type %T has no httprequest.Route field", params)
+	}
+	reqURL := baseURL
+	if rt.path != "" {
+		reqURL += rt.path
+	}
+	req, err := Marshal(reqURL, rt.method, params)
+	if err != nil {
+		return errgo.Mask(err)
+	}
+
+	// Actually make the request.
+	doer := c.Doer
+	if doer == nil {
+		doer = http.DefaultClient
+	}
+	var httpResp *http.Response
+	body := req.Body.(BytesReaderCloser)
+	if doer1, ok := doer.(DoerWithBody); ok && body.Len() > 0 {
+		req.Body = nil
+		httpResp, err = doer1.DoWithBody(req, body)
+	} else {
+		httpResp, err = doer.Do(req)
+	}
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+
+	// Return response directly if required.
+	if respPt, ok := resp.(**http.Response); ok {
+		*respPt = httpResp
+		return nil
+	}
+	defer httpResp.Body.Close()
+	if httpResp.StatusCode == http.StatusOK {
+		return unmarshalJSONResponse(httpResp, resp)
+	}
+
+	errUnmarshaler := c.UnmarshalError
+	if errUnmarshaler == nil {
+		errUnmarshaler = DefaultErrorUnmarshaler
+	}
+	err = errUnmarshaler(httpResp)
+	if err == nil {
+		err = errgo.Newf("unexpected HTTP response status: %s", httpResp.Status)
+	}
+	return err
+}
+
+// ErrorUnmarshaler returns a function which will unmarshal error
+// responses into new values of the same type as template. The argument
+// v must be a pointer. A new instance of it is created every time the
+// returned function is called.
+func ErrorUnmarshaler(template error) func(*http.Response) error {
+	t := reflect.TypeOf(template)
+	if t.Kind() != reflect.Ptr {
+		panic(errgo.Newf("cannot unmarshal errors into value of type %T", template))
+	}
+	t = t.Elem()
+	return func(resp *http.Response) error {
+		if 300 <= resp.StatusCode && resp.StatusCode < 400 {
+			// It's a redirection error.
+			loc, _ := resp.Location()
+			return fmt.Errorf("unexpected redirect (status %s) from %q to %q", resp.Status, resp.Request.URL, loc)
+		}
+		if err := checkIsJSON(resp.Header); err != nil {
```

Comment:
We do see a lot of not JSON errors, it would be nice if they did something vaguely sensible. A RemoteError with the Content-Type and Body in the info parameter is one suggestion. It would be better not to lose the context from the body.